flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/ExecutionSlotAllocatorFactory.java

```diff
@@ -19,11 +19,39 @@
 
 package org.apache.flink.runtime.scheduler;
 
+import org.apache.flink.runtime.clusterframework.types.AllocationID;
+import org.apache.flink.runtime.clusterframework.types.ResourceProfile;
+import org.apache.flink.runtime.jobmanager.scheduler.CoLocationGroupDesc;
+import org.apache.flink.runtime.jobmanager.scheduler.SlotSharingGroup;
+import org.apache.flink.runtime.scheduler.strategy.ExecutionVertexID;
+import org.apache.flink.runtime.scheduler.strategy.SchedulingTopology;
+
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
 /**
  * Interface for {@link ExecutionSlotAllocator} factories.
  */
+@FunctionalInterface
 public interface ExecutionSlotAllocatorFactory {
 
-	ExecutionSlotAllocator createInstance(PreferredLocationsRetriever preferredLocationsRetriever);
-
+	/**
+	 * Instantiates the {@link ExecutionSlotAllocator}.
+	 *
+	 * @param preferredLocationsRetriever to retrieve preferred locations for each execution
+	 * @param resourceProfileRetriever to retrieve required {@link ResourceProfile} for each execution
+	 * @param priorAllocationIdRetriever to retrieve prior {@link AllocationID} for each execution
+	 * @param topology scheduling topology of the job
+	 * @param logicalSlotSharingGroupSupplier to supply all slot sharing groups in the job
+	 * @param coLocationGroupSupplier to supply all co-location groups in the job
+	 * @return The instantiated slot allocator
+	 */
+	ExecutionSlotAllocator createInstance(
+		PreferredLocationsRetriever preferredLocationsRetriever,
+		Function<ExecutionVertexID, ResourceProfile> resourceProfileRetriever,
+		Function<ExecutionVertexID, AllocationID> priorAllocationIdRetriever,
+		SchedulingTopology topology,
+		Supplier<Set<SlotSharingGroup>> logicalSlotSharingGroupSupplier,
+		Supplier<Set<CoLocationGroupDesc>> coLocationGroupSupplier);
```

Comment:
Making it suppliers make it possible to invoke `getSlotSharingGroups()` and `getCoLocationGroupDescriptors()` lazily.
It's better to not invoke them if it is not `SlotSharingExecutionSlotAllocator` .

Reply:
One example is that if a user uses eager scheduling, `getSlotSharingGroups()` and `getCoLocationGroupDescriptors()` will not be invoked if we use supplier here because they will not be used.

Reply:
The `ExecutionSlotAllocationContext` is reworked so we do not need `Supplier` anymore.